### PR TITLE
chore: Update NPM dependencies

### PR DESCRIPTION
**Pull request by mjameswh**
## What changed

- Updated NPM dependencies. A lot of them.
- Pruned left-over unused dependencies
- Added Node 18 and 20 to CI matrix for build-and-test
- Minimal reformat of source files due to updating prettier
- Updated Core SDK to a764405
- Added maintenance instructions to CONTRIBUTING.md

## What did _not_ change

The following dependencies were NOT updated:

- **ava** (current 4.3.3, latest 6.0.1): ava 5 requires Node 14.19+ or 16.15+; ava 6.x requires Node 18+.
- **lerna** (current 6.6.2, latest 8.0.2): lerna 7.x requires Node 16+, and lerna 8.x requires Node 18+.
- **@grpc/grpc-js** (current 1.7.3, latest 1.9.13): updating would probably be safe at this point, but will deal with this one in a different PR, as it requires special testing to guarantee that the new lib-side retry policy introduced in 1.8.0 is not conflicting with our own retry logic.
- **got** (current 13.0.0, latest 14.0.0): got 14.x requires Node 20.
- **node-fetch** (current 2.7.0, latest 3.3.2): node-fetch 3.x is ESM only

### CODE REVIEW

**Review by bergundy — APPROVED** [review]
Thanks!

Can you please mention how you updated these in the PR description? What command did you run?

File: .eslintrc [hunk 1]
```diff
@@ -38,8 +38,6 @@
     "import/no-relative-packages": "error",
     "import/export": "error",
     "import/no-named-as-default": "error",
-    // TODO enable and fix
-    // "import/no-deprecated": "error",
```
Comment (mjameswh) [inline .eslintrc:42]: `import/no-deprecated` causes lot of false-positive. Actually, it gave nothing pertinent compared to `deprecation/deprecation`.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -143,6 +143,43 @@ chore(samples): upgrade commander module
 
 The `scope` options are listed in [commitlint.config.js](https://github.com/temporalio/sdk-typescript/blob/main/commitlint.config.js).
 
+## Updating and pruning dependencies
+
+There are various tools out there to help with updating and pruning NPM dependencies. Unfortunately,
+most of these tools don't work well in older lerna-style monorepos, such as this one, and updating
+to NPM 7 workspace would break support for Node 14.x. We must therefore rely on a more manual
```
Comment (bergundy) [inline CONTRIBUTING.md:150]: I'd definitely support dropping support for node 14 in the next minor release.

**Review by bergundy — APPROVED** [review]

**Comment by mjameswh** [comment]
Forcing merge as the list of required checks was outdated. I just reconfigured the project's requirements, but refreshing this PR will result in a new round of GHA execution, which I want to avoid at this point.

**Pull request merged**